packages/runtime-core/src/componentProps.ts

```diff
@@ -19,24 +19,30 @@ import { Data, ComponentInternalInstance } from './component'
 
 export type ComponentPropsOptions<P = Data> =
   | ComponentObjectPropsOptions<P>
-  | string[]
+  | Extract<keyof P, string>[]
 
 export type ComponentObjectPropsOptions<P = Data> = {
   [K in keyof P]: Prop<P[K]> | null
 }
 
 export type Prop<T> = PropOptions<T> | PropType<T>
 
+type DefaultFactory<T> = () => T | null | undefined
+
 interface PropOptions<T = any> {
   type?: PropType<T> | true | null
   required?: boolean
-  default?: T | null | undefined | (() => T | null | undefined)
-  validator?(value: unknown): boolean
+  default?:
+    | DefaultFactory<T>
+    | null
+    | undefined
+    | (T extends Function ? never : T) // If type of prop is function, only allow factory function
```

Comment:
I'm not sure if this makes sense. Actually it seems to break certain cases.

Reply:
Which ones?